master/buildbot/process/build.py

```diff
@@ -284,30 +286,14 @@ def startBuild(self, build_status, workerforbuilder):
             # handler
             log.msg("Build.setupBuild failed")
             log.err(Failure())
-            self.finished = True
-            self.results = EXCEPTION
-            self.deferred = None
+            self.buildFinished(['Build.setupBuild', 'failed'], EXCEPTION)
             return
         # flush properties in the beginning of the build
         yield self._flushProperties(None)
-        yield self.master.data.updates.setBuildStateString(self.buildid,
-                                                           u'starting')
-        self.build_status.buildStarted(self)
         yield self.acquireLocks()
 
-        try:
-            # start the sequence of steps and wait until it's finished
-            self.startNextStep()
-            yield self.deferred
```

Comment:
This is only ever fired by `Build.buildFinished`. Instead of doing some cleanup there and some here, just do it all there and don't bother waiting.